File: Source/Java/RiceDoctor/src/main/java/vn/edu/uit/ontologymanager/model/Relation.java
package vn.edu.uit.ontologymanager.model;

import com.google.gson.annotations.SerializedName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import vn.edu.uit.shared.Buildable;

import java.util.Set;

import static vn.edu.uit.ontologymanager.model.EntityType.RELATION;

public class Relation extends Entity {

    @Nullable
    @SerializedName("DirectDomains")
    private final Set<Class> directDomains;

    @Nullable
    @SerializedName("AllDomains")
    private final Set<Class> allDomains;

    @Nullable
    @SerializedName("DirectRanges")
    private final Set<Class> directRanges;

    @Nullable
    @SerializedName("AllRanges")
    private final Set<Class> allRanges;

    Relation(@NotNull final Builder builder) {
        super(RELATION, builder.id, builder.label);
        directDomains = builder.directDomains;
        allDomains = builder.allDomains;
        directRanges = builder.directRanges;
        allRanges = builder.allRanges;
    }

    public static class Builder implements Buildable<Relation> {

        @NotNull
        private final String id;

        @Nullable
        private final String label;

        @Nullable
        private Set<Class> directDomains;

        @Nullable
        private Set<Class> allDomains;

        @Nullable
        private Set<Class> directRanges;

        @Nullable
        private Set<Class> allRanges;

        public Builder(@NotNull final String id, @Nullable final String label) {
            this.id = id;
            this.label = label;
        }

        @NotNull
        public Builder directDomains(@Nullable final Set<Class> directDomains) {
            this.directDomains = directDomains;
            return this;
        }

        @NotNull
        public Builder allDomains(@Nullable final Set<Class> allDomains) {
            this.allDomains = allDomains;
            return this;
        }

        @NotNull
        public Builder directRanges(@Nullable final Set<Class> directRanges) {
            this.directRanges = directRanges;
            return this;
        }

        @NotNull
        public Builder allRanges(@Nullable final Set<Class> allRanges) {
            this.allRanges = allRanges;
            return this;
        }

        @Override
        @NotNull
        public Relation build() {
            return new Relation(this);
        }
    }
}

File: Source/C#/RiceDoctor/RiceDoctor.OntologyManager/Models/Relation.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;
using RiceDoctor.Shared;
using static RiceDoctor.OntologyManager.GetType;

namespace RiceDoctor.OntologyManager
{
    public class Relation : Entity<Relation>, IEquatable<Relation>
    {
        [CanBeNull] private IReadOnlyCollection<Class> _allDomains;
        [CanBeNull] private IReadOnlyCollection<Class> _allRanges;
        private bool _canGetAllDomains;
        private bool _canGetAllRanges;
        private bool _canGetDirectDomains;
        private bool _canGetDirectRanges;
        private bool _canGetInverseRelation;
        [CanBeNull] private IReadOnlyCollection<Class> _directDomains;
        [CanBeNull] private IReadOnlyCollection<Class> _directRanges;
        [CanBeNull] private Relation _inverseRelation;

        public Relation([NotNull] string id, [CanBeNull] string label = null) : base(id, label)
        {
        }

        public override EntityType Type => EntityType.Relation;

        public bool Equals(Relation other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return base.Equals(other);
        }

        [CanBeNull]
        public Relation GetInverseRelation()
        {
            if (_canGetInverseRelation) return _inverseRelation;

            _inverseRelation = Manager.Instance.GetInverseRelation(Id);
            _canGetInverseRelation = true;

            return _inverseRelation;
        }

        public void SetInverseRelation([CanBeNull] Relation inverseRelation)
        {
            if (_canGetInverseRelation)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetInverseRelation)));

            _inverseRelation = inverseRelation;
            _canGetInverseRelation = true;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetDirectDomains()
        {
            if (_canGetDirectDomains) return _directDomains;

            _directDomains = Manager.Instance.GetRelationDomains(Id, GetDirect);
            _canGetDirectDomains = true;

            return _directDomains;
        }

        public void SetDirectDomains([CanBeNull] IReadOnlyCollection<Class> directDomains)
        {
            if (_canGetDirectDomains)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetDirectDomains)));

            _directDomains = directDomains;
            _canGetDirectDomains = true;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetAllDomains()
        {
            if (_canGetAllDomains) return _allDomains;

            _allDomains = Manager.Instance.GetRelationDomains(Id, GetAll);
            _canGetAllDomains = true;

            return _allDomains;
        }

        public void SetAllDomains([CanBeNull] IReadOnlyCollection<Class> allDomains)
        {
            if (_canGetAllDomains)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetAllDomains)));

            _allDomains = allDomains;
            _canGetAllDomains = true;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetDirectRanges()
        {
            if (_canGetDirectRanges) return _directRanges;

            _directRanges = Manager.Instance.GetRelationRanges(Id, GetDirect);
            _canGetDirectRanges = true;

            return _directRanges;
        }

        public void SetDirectRanges([CanBeNull] IReadOnlyCollection<Class> directRanges)
        {
            if (_canGetDirectRanges)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetDirectRanges)));

            _directRanges = directRanges;
            _canGetDirectRanges = true;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetAllRanges()
        {
            if (_canGetAllRanges) return _allRanges;

            _allRanges = Manager.Instance.GetRelationRanges(Id, GetAll);
            _canGetAllRanges = true;

            return _allRanges;
        }

        public void SetAllRanges([CanBeNull] IReadOnlyCollection<Class> allRanges)
        {
            if (_canGetAllRanges)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetAllRanges)));

            _allRanges = allRanges;
            _canGetAllRanges = true;
        }

        public static bool operator ==(Relation relation1, Relation relation2)
        {
            if (ReferenceEquals(relation1, relation2)) return true;
            if (ReferenceEquals(null, relation1)) return false;
            if (ReferenceEquals(null, relation2)) return false;
            return relation1.Equals(relation2);
        }

        public static bool operator !=(Relation relation1, Relation relation2)
        {
            return !(relation1 == relation2);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Relation) obj);
        }

        public override int GetHashCode()
        {
            return base.GetHashCode();
        }
    }
}